### PR TITLE
[ACPBoard #378] Bugfix 500 Server Error deleting missing identities

### PR DESCRIPTION
**Pull request by KashifSaadat**
- Added additional route to show identity, returning a json object
- Added 'before_action' on delete and show identity, to first attempt a get and return 404 on failure
- Updated controller tests

### CODE REVIEW

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 1]
```diff
@@ -42,6 +42,87 @@
     end
   end
 
+  describe '#show_identity' do
+    it_behaves_like 'authenticated' do
+      before do
+        current_user.identities.create!(
```
Comment (marcinc) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: use factory here instead:
```ruby
create(:identity, 
  external_id: '1234567',
  external_username: test_auth_payload.preferred_username,
  ...
)
```

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 2]
```diff
@@ -42,6 +42,87 @@
...
+        expect(json_response['error']['message']).to eq "No such identity 'kubernetes' currently exists for this user"
+        expect(Audit.count).to eq 0
+      end
+    end
+  end
+
+  describe '#delete_identity' do
+    it_behaves_like 'authenticated' do
+      before do
+        current_user.identities.create!(
```
Comment (marcinc) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: Same here, best to use factory.

File: platform-hub-api/app/controllers/me_controller.rb [hunk 3]
```diff
@@ -12,29 +10,23 @@ def show
...
-      logger.warn "User '#{current_user.email}' attempted to access an identity which does not exist: #{params[:service]}"
-      render_error "No such identity '#{params[:service]}' currently exists for this user", :not_found
-    end
-  end
-
   def delete_identity
     # We assume that the `service` param has been validated by the route constraints
-    @user_identity.destroy
+    user_identity = current_user.identity(params[:service])
+    if !user_identity.nil?
```
Comment (marcinc) [inline platform-hub-api/app/controllers/me_controller.rb]: you could also express it in slightly more idiomatic way
```ruby
if user_identity.present?
#or 
unless user_identity.nil?
```

File: platform-hub-api/app/controllers/me_controller.rb [hunk 4]
```diff
@@ -12,15 +12,20 @@ def show
 
   def delete_identity
     # We assume that the `service` param has been validated by the route constraints
-    current_user.identity(params[:service]).destroy
+    user_identity = current_user.identity(params[:service])
+    if !user_identity.nil?
```
Comment (jits) [inline platform-hub-api/app/controllers/me_controller.rb]: `if user_identity.present?` is a slightly nicer way to write this.

File: platform-hub-api/app/controllers/me_controller.rb (hunk 3 above)
Comment (marcinc) [inline platform-hub-api/app/controllers/me_controller.rb]: Alternatively you can call destroy on `user_identity` without checking for presence and rescue `ActiveRecord::RecordNotFound` exception.

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 5]
```diff
@@ -42,6 +42,51 @@
...
+          external_id: '1234567',
+          external_username: test_auth_payload.preferred_username,
+          external_name: test_auth_payload.name,
+          external_email: test_auth_payload.email
+        )
+      end
+
+      it 'should delete an identity successfully if it exists for the current user' do
+        expect(Audit.count).to eq 0
+        delete :delete_identity, params: {service: 'github'}
```
Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: The provider `github` here just happens to match the provider of the factory object you created above; a potentially nicer way would be to do `{service: @identity.provider}` here and make sure you assign `@identity = <factory object creation>` above.

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 6]
```diff
@@ -42,6 +42,51 @@
     end
   end
 
+  describe '#delete_identity' do
+    it_behaves_like 'authenticated' do
+      before do
+        create(:identity,
+          user: current_user,
+          external_id: '1234567',
```
Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: For the purpose of this test, you probably don't need to explicitly set this field and all the ones below it.

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 7]
```diff
@@ -42,6 +42,51 @@
...
+        expect(Audit.count).to eq 0
+        delete :delete_identity, params: {service: 'github'}
+        expect(response).to be_success
+        expect(current_user.identity('github')).to be nil
+        expect(Audit.count).to eq 1
+        expect(Audit.first.action).to eq 'delete_identity'
+        expect(Audit.first.user.id).to eq current_user_id
+      end
+
+      it 'should gracefully fail to delete an identity that does not exist' do
```
Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: If you don't mind, could you restructure these test cases to be more in line with how we do 404 test cases elsewhere ([example](https://github.com/UKHomeOffice/platform-hub/blob/master/platform-hub-api/spec/controllers/projects_controller_spec.rb#L51-L73)).

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 8]
```diff
@@ -42,6 +42,51 @@
...
+
+      it 'should gracefully fail to delete an identity if called twice' do
+        expect(Audit.count).to eq 0
+        delete :delete_identity, params: {service: 'github'}
+        expect(response).to be_success
+        expect(current_user.identity('github')).to be nil
+        expect(Audit.count).to eq 1
+        expect(Audit.first.action).to eq 'delete_identity'
+        expect(Audit.first.user.id).to eq current_user_id
+        delete :delete_identity, params: {service: 'github'}
```
Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: Likewise here, and the two other usages above – should use `@identity.provider` to couple the value to the actual factory object.

**Comment by KashifSaadat** [comment]
Updated following comments, rebased off master and squished additional commits.

File: platform-hub-api/spec/controllers/me_controller_spec.rb [hunk 9]
```diff
@@ -42,6 +42,37 @@
...
+          expect(current_user.identity(@identity.provider)).to be nil
+          expect(Audit.count).to eq 1
+          expect(Audit.first.action).to eq 'delete_identity'
+          expect(Audit.first.user.id).to eq current_user_id
+        end
+      end
+
+      context 'for a non-existent identity' do
+        it 'should fail and return a 404' do
+          delete :delete_identity, params: {service: 'kubernetes'}
```
Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: Just to make it super obvious, could do something like `service: 'unknown_service'`.

Comment (jits) [inline platform-hub-api/spec/controllers/me_controller_spec.rb]: (discussed offline – it's tricky to do what I suggested due this being a route level constraint, so okay as-is).

**Pull request merged**